lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,296 @@
+        task_name = self.rtapp_task_ids_map['task'][0].comm
+
+        df = self.trace.analysis.rta.df_phases(task_name)
+        df = df.reset_index()
+        df.rename(columns={'index': 'start'}, inplace=True)
+        df['end'] = df['start'].shift(-1)
+        df = df[df.phase > 0]
+
+        df[['capacity', 'capacity_max']] = df.apply(
+                lambda x:
+                self._get_task_activation_capacity(task_name, x['start'],
+                                                   x['end']),
+                axis=1, result_type='expand')
+
+        phases = self._get_phases(self.plat_info)
+        phases.insert(0, 0)
+
+        df[['uclamp_val', 'util', 'band']] = df.apply(
+                lambda x: [phases[int(x.phase)][0],
+                           phases[int(x.phase)][1],
+                           phases[int(x.phase)][2]],
+                axis=1, result_type='expand')
+
+        return df
+
+    @classmethod
+    def _test_passed(cls, df, col, minmax_func):
+        df[['expected_min', 'expected_max']] = df.apply(
+                lambda x: minmax_func(x),
+                axis=1, result_type='expand')
+
+        df['pass'] = ((
+            df[col] >= df['expected_min']) &
+            (df[col] <= df['expected_max']))
+
+        print(df)
+
+        return ResultBundle.from_bool(df['pass'].all())
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_placement(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement matches UtilClamp
+        requirements.
+
+        The expected capacity is the one available just above the UtilClamp
+        value +/- 10% (to handle migration time)
+        """
+
+        df = self._get_analysis()
+        max_capacities = sorted(list(dict.fromkeys(
+            self.plat_info['cpu-capacities']['rtapp'].values())))
+
+        def _expected_minmax(row):
+            uclamp_val = row['uclamp_val']
+            expected = 1024
+
+            for cap in max_capacities:
+                # kernel task placement uses 80% margin on the capacity
+                if (cap * 0.8) > uclamp_val:
+                    expected = cap
+                    break
+
+            margin = int(expected * 0.1)
+
+            return [expected - margin, expected + margin]
+
+        return self._test_passed(df, 'capacity_max', _expected_minmax)
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_freq_selection(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement and frequency selection
+        match UtilClamp requirements.
+
+        The expected capacity is :
+          - min: The first capacity below the UtilClamp value.
+          - max: The 4th capacity above the UtilClamp value. This arbitrary
+            value must be high enough to cover the migration and schedutil
+            margins.
```

Comment:
Same as for the other test, we don't want arbitrary margins. What is are the issues exactly when we don't have such margin ? We need to know precisely, and then take them into account (schedutil is fairly easy to simulate since we have access to util_est).

Reply:
I don't think we can easily predict the schedutil selected frequency:
* util_est doesn't use the clamped value of util.
* the usage of the uclamp value for schedutil depends _when_ schedutil updates the frequency. (i.e. If it is updated while the task is enqueued or not)
* other things running on the board that would change the schedutil value, including bits that aren't frozen during the run, like IRQs.

Notice also, that this max value is more a sanity check that the solution to a UtilClamp value isn't just using the highest capacity available. Ultimately, we only test the min value of UClamp.

I can try to compute what would be the outcome of get_next_freq() for a given util, but that would mean doing many assumptions anyway.

Reply:
> util_est doesn't use the clamped value of util.

You can just use that:
https://lisa-linux-integrated-system-analysis.readthedocs.io/en/master/trace_analysis.html#lisa.analysis.load_tracking.LoadTrackingAnalysis.df_task_signal

> the usage of the uclamp value for schedutil depends when schedutil updates the frequency. (i.e. If it is updated while the task is enqueued or not)
We have `cpu_frequency` event to tell us when schedutil took a decision to change freq.

> other things running on the board that would change the schedutil value, including bits that aren't frozen during the run, like IRQs.
We have a couple of options:
* simulate schedutil behavior on each kind of util so we know what to expect for a given CFS task util (`total/1024 * max_freq` plus checking that no iowait boost or similar occured. If they did, we can just ignore the "unpredictable" sample) 
* make schedutil a bit more verbose on what it does and why.

The alternative is simply only check the placement with uclamp, and not the frequency selection. It might look artificial and "why don't we do it since we can ?" but in reality, if we cannot implement what I described, we cannot actually check what is happening. Tests relying on margins as a cheap way of "simulating" kernel subsystems have all turned into maintenance liability, and some of them have empirically been shown to not be able to even catch using the energy model description of another board than the one under test.

The only sort-of-exception to that are tests checking one-sided interval. If for example we should *never* end up selecting a frequency below a threshold (taking into account all the edge conditions), we can check for that, and this property should hold across all platforms.